.github/workflows/pr-checks.yml

```diff
@@ -34,3 +34,24 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - run: echo ${{ matrix.directory }}
+
+  terraform-plan:
+    name: Terraform Plan
+    needs: terraform-changes
+    strategy:
+      matrix:
+        directory: ${{ fromJSON(needs.terraform-changes.outputs.directories).matrix }}
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: hashicorp/setup-terraform@v2
+      - env:
+          AWS_DEFAULT_REGION: "eu-west-2"
+          GITHUB_TOKEN: ${{ secrets.GH_ACCESS_TOKEN }}
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_STATE_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_STATE_SECRET_ACCESS_KEY }}
+          DIRECTORY: ${{ matrix.directory }}
+        run: |
+          cd terraform/${DIRECTORY}
+          terraform init -backend-config="../../backend.hcl" -backend-config="key=${DIRECTORY}/terraform.tfstate"
```

Comment:
I thought enforcing the key in the CI/CD to be created via the directory name would ensure there's minimal chance of key duplication - we probably need to think about what happens if a folder gets deleted and recreated 🤔

Reply:
Maybe you could use a tool like hcl2json to parse all the keys on PR and reject if there’s a duplicate

Reply:
I was looking reading up about it last night - I don't believe it'd be suitable for. the template since it seemed like the community was fairly small i.e. not a lot of eyeballs to spot bad actors 👀 and could potentially fall flat, so didn't feel like something I'd trust enough to build dependency on it if we can do without

I'm happy to give it a whirl, but I'd probably want to step through the code first and make sure it's suitable and since it's written in GO - I'd need help with that 😂 

Anything with a smallish community, we probably want to ensure we have the skills and knowledge to fork it and carry it forward if it became unsupported

Reply:
That being said, my analysis was pretty much just looking at the stars on Github - so maybe the community is actually a lot larger than I think!

Reply:
Maybe https://pypi.org/project/python-hcl2/ .

Reply:
Didn't spot that one, but shows there's some appetite for it! 🍔 

From doing some reading last night, it sounded like there's not an officially supported way to convert HCL into JSON nor did it seem like there would be in the near future.

So the concern is when HCL3 comes out and there's loads of new language features we want to use, but can't because we're dependent on a parser that only works for HCL2 

So I think for the time being, using the folder per key convention meets the need (with a few edge cases I need to look into) without bringing in a potentially brittle third-party tool. Though if those edge cases become too complex, I'd prefer to go with a third party tool and see how it goes 😂 all about finding the tradeoffs ⚖️ 

Happy to discuss this more, and it will need to be periodically reviewed! Since there's no point in us supporting a new wheel if we can bring something tried and tested into the workflow instead 😁

Reply:
But doing some more reading now and just found this [https://developer.hashicorp.com/terraform/cdktf/concepts/hcl-interoperability](https://developer.hashicorp.com/terraform/cdktf/concepts/hcl-interoperability )👀

Reply:
Sounds like another pulumi, not what I thought 😂

Reply:
Beat me to it!

Reply:
In my experience, I’ve never run into duplicate keys. They should be caught by a plan, because you’ll see Terraform trying to tear down something as it tries to reconcile what you’ve asked for in that current folder

Reply:
Hiya :wave: This appears to be an issue only because there's concern about duplicates. And the duplicates aren't a huge concern if we don't insist on reusing the backend settings. Using a single backend file for multiple projects is not a good pattern for vanilla terraform. For teams that do want to go down this road, a secondary tool is recommended, like terragrunt. Also, how far down in a directory tree do you go? Even in most monorepos that I've seen, terraform is typically set up as something like `terraform/project` and another level for internal modules. Anything more than that is not something that should be encouraged or accounted for, in my opinion, as that is not a good pattern.